src/NUnitFramework/framework/Constraints/UniqueItemsConstraint.cs

```diff
@@ -91,58 +129,50 @@ private bool OriginalAlgorithm(IEnumerable actual)
                     return CharsUniqueIgnoringCase((IEnumerable<char>)actual);
             }
 
-            return (bool)ItemsUniqueMethod.MakeGenericMethod(memberType).Invoke(null, new object[] { actual });
+            return (ICollection)ItemsUniqueMethod.MakeGenericMethod(memberType).Invoke(null, new object[] { actual });
         }
 
-        private bool IsSealed(Type type)
+        private static bool IsSealed(Type type)
         {
             return type.GetTypeInfo().IsSealed;
         }
 
         private static readonly MethodInfo ItemsUniqueMethod =
             typeof(UniqueItemsConstraint).GetMethod(nameof(ItemsUnique), BindingFlags.Static | BindingFlags.NonPublic);
 
-        private static bool ItemsUnique<T>(IEnumerable<T> actual)
-        {
-            var hash = new HashSet<T>();
-
-            foreach (T item in actual)
-            {
-                if (!hash.Add(item))
-                    return false;
-            }
-
-            return true;
-        }
-
-        private static bool StringsUniqueIgnoringCase(IEnumerable<string> actual)
-        {
-            var hash = new HashSet<string>();
-
-            foreach (string item in actual)
-            {
-                string s = item.ToLower();
+        private static ICollection ItemsUnique<T>(IEnumerable<T> actual)
+            => NonUniqueItemsInternal(actual, v => v);
 
-                if (!hash.Add(s))
-                    return false;
-            }
+        private static ICollection StringsUniqueIgnoringCase(IEnumerable<string> actual)
+            => NonUniqueItemsInternal(actual, v => v.ToLower());
 
-            return true;
-        }
+        private static ICollection CharsUniqueIgnoringCase(IEnumerable<char> actual)
+            => NonUniqueItemsInternal(actual, char.ToLower);
 
-        private static bool CharsUniqueIgnoringCase(IEnumerable<char> actual)
+        private static ICollection NonUniqueItemsInternal<T>(IEnumerable<T> actual, Func<T,T> hashValueFactory)
```

Comment:
We can't pass a `IEqualityComparer<T>` in here and use that when constructing the `Dictionary<T,int> below, as the linux tests seem to fail when uniquely detecting case insensitive strings. The only way I could get that to work was to first do a `ToLower()` and storing that in the dictionary (this was also what the previous version of the code did).

Reply:
I guess the `ToLower()` will give us problems in some locale, but this was the same issue before. Is the test failure the same kind as we corrected in #3416, meaning that they could be solved by annotating the tests?